dhcpd.8
.\"	dhcpd.8
.\"
.\" Copyright (c) 1995, 1996 The Internet Software Consortium.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of The Internet Software Consortium nor the names
.\"    of its contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE INTERNET SOFTWARE CONSORTIUM AND
.\" CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL THE INTERNET SOFTWARE CONSORTIUM OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
.\" USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
.\" OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" This software has been written for the Internet Software Consortium
.\" by Ted Lemon <mellon@fugue.com> in cooperation with Vixie
.\" Enterprises.  To learn more about the Internet Software Consortium,
.\" see ``http://www.isc.org/isc''.  To learn more about Vixie
.\" Enterprises, see ``http://www.vix.com''.
.TH dhcpd 8
.SH NAME
dhcpd - Dynamic Host Configuration Protocol server
.SH SYNOPSIS
.B dhcpd
[
.B -p
.I port
]
[
.B -f
]
[
.I if0
[
.I ...ifN
]
]
.SH DESCRIPTION
dhcpd(8) implements the Dynamic Host Configuration Protocol (DHCP) and
the Internet Bootstrap Protocol (BOOTP).  DHCP allows hosts on a
TCP/IP network to request and be assigned IP addresses, and also to
discover information about the network to which they are attached.
BOOTP provides similar but much more limited functionality.
.SH OPERATION
.PP
The DHCP protocol allows a host which is unknown to the network
administrator to be automatically assigned a new IP address out of a
pool of IP addresses for its network.   In order for this to work, the
network administrator allocates address pools in each subnet and
enters them into the dhcpd.conf(5) file.
.PP
On startup, dhcpd reads the
.IR dhcpd.conf
file and keeps the list of available addresses on each subnet in
memory.  When a host requests an address using the DHCP protocol,
dhcpd allocates an address for it.  Each such host is assigned a
lease, which expires after an amount of time chosen by the
administrator (by default, one day).  As leases expire, the hosts to
which they are assigned are expected to renew the leases if they wish
to continue to use the addresses.   Once a lease has expired, the host
to which that lease is assigned is no longer permitted to use the IP
address assigned to it.
.PP
In order to keep track of leases across system reboots and server
restarts, dhcpd keeps a list of leases it has assigned in the
dhcpd.leases(5) file.   Before dhcpd grants a lease to a host, it
records the lease in this file and makes sure that the contents of the
file are flushed to disk.   This ensures that even in the event of a
system crash, dhcpd will not forget about a lease that it has
assigned.   On startup, after reading the dhcpd.conf file, dhcpd
reads the dhcpd.leases file to refresh its memory about what leases
have been assigned.
.PP
New leases are appended to the end of the dhcpd.leases
file.   In order to prevent the file from becoming arbitrarily large,
from time to time dhcpd creates a new dhcpd.leases file from its
in-core lease database.  Once this file has been written to disk, the
old file is renamed
.IR dhcpd.leases~ ,
and the new file is renamed dhcpd.leases.   If the system crashes in
the middle of this process, whichever dhcpd.leases file remains will
contain all the lease information, so there is no need for a special
crash recovery process.
.PP
BOOTP support is also provided by this server.   Unlike DHCP, the
BOOTP protocol requires that the server know the hardware address of
the client that is to be booted.   The network administrator must
determine that address, allocate an IP address for the client, and
enter that information into the dhcpd.conf file.
.PP
Whenever changes are made to the dhcpd.conf file, dhcpd must be
restarted.   To restart dhcpd, send a SIGTERM (signal 15) to the
process ID contained in
.IR /dhcpd.pid ,
and then re-invoke dhcpd.

.SH COMMAND LINE
.PP
dhcpd normally identifies all interfaces on the system which are up,
and listens on each interface.   If possible, point-to-point
interfaces and the loopback interface are eliminated, but on some
systems this is not possible.   For this reason, the interfaces on
which dhcp should listen may be explicitly specified on the command
line.
.PP
dhcpd normally listens on port 67, which is the BOOTP Server Port
(the DHCP and BOOTP protocols both use this port).   If desired, dhcpd
may be invoked with the
.B -p
flag, followed by a port number, so as to provide DHCP service on a
different port.   This is mostly useful for debugging purposes.
.PP
On some System-V systems, it may be desirable to run dhcp from
/etc/inittab.   If so, dhcpd should be invoked with the
.B -f
flag, which causes dhcpd to run in the foreground; otherwise, dhcpd
automatically detaches itself from the process group that started it
and runs in the background.
.SH CONFIGURATION
The syntax of the dhcpd.conf(8) file is discussed seperately.   This
section should be used as an overview of the configuration process,
and the dhcpd.conf(8) documentation should be consulted for detailed
reference information.
.PP
.SH Subnets
dhcpd needs to know the subnet numbers and netmasks of all subnets for
which it will be providing service.   In addition, in order to
dynamically allocate addresses, it must be assigned one or more ranges
of addresses on each subnet which it can in turn assign to client
hosts as they boot.   Thus, a very simple configuration providing DHCP
support might look like this:
.nf
.sp 1
	subnet 239.252.197.0 netmask 255.255.255.0
	  range 239.252.197.10 239.252.197.250;
.fi
.PP
Multiple address ranges may be specified like this:
.nf
.sp 1
	subnet 239.252.197.0 netmask 255.255.255.0
	  range 239.252.197.10 239.252.197.107
	  range 239.252.197.113 239.252.197.250;
.fi
.PP
If a subnet will only be provided with BOOTP service and no dynamic
address assignment, the range clause can be left out entirely, but the
subnet statement must appear.
.PP
.SH Lease Lengths
DHCP leases can be assigned almost any length from zero seconds to
infinity.   What lease length makes sense for any given subnet, or for
any given installation, will vary depending on the kinds of hosts
being served.
.PP
For example, in an office environment where systems are added from
time to time and removed from time to time, but move relatively
infrequently, it might make sense to allow lease times of a month of
more.   In a final test environment on a manufacturing floor, it may
make more sense to assign a maximum lease length of 30 minutes -
enough time to go through a simple test procedure on a network
appliance before packaging it up for delivery.
.PP
It is possible to specify two lease lengths: the default length that
will be assigned if a client doesn't ask for any particular lease
length, and a maximum lease length.   These are specified as clauses
to the subnet command:
.nf
.sp 1
	subnet 239.252.197.0 netmask 255.255.255.0
	  range 239.252.197.10 239.252.197.107
	  default-lease-time 600
	  max-lease-time 7200;
.fi
.PP
This particular subnet declaration specifies a default lease time of
600 seconds (ten minutes), and a maximum lease time of 7200 seconds
(two hours).   Other common values would be 86400 (one day), 604800
(one week) and 2592000 (30 days).
.PP
Each subnet need not have the same lease\(emin the case of an office
environment and a manufacturing environment served by the same DHCP
server, it might make sense to have widely disparate values for
default and maximum lease times on each subnet.
.SH BOOTP Support
Each BOOTP client must be explicitly declared in the dhcpd.conf
file.   A very basic client declaration will specify the client
network interface's hardware address and the IP address to assign to
that client.   If the client needs to be able to load a boot file from
the server, that file's name must be specified.   A simple bootp
client declaration might look like this:
.nf
.sp 1
	host haagen hardware ethernet 08:00:2b:4c:59:23
	  fixed-address 239.252.197.9
	  filename "/tftpboot/haagen.boot";
.fi
.SH Options
DHCP (and also BOOTP with Vendor Extensions) provide a mechanism
whereby the server can provide the client with information about how
to configure its network interface (e.g., subnet mask), and also how
the client can access various network services (e.g., DNS, IP routers,
and so on).
.PP
These options can be specified on a per-subnet basis, and, for BOOTP
clients, also on a per-client basis.   In the event that a BOOTP
client declaration specifies options that are also specified in its
subnet declaration, the options specified in the client declaration
take precedence.   An reasonably complete DHCP configuration might
look something like this:
.nf
.sp 1
	subnet 239.252.197.0 netmask 255.255.255.0
	  range 239.252.197.10 239.252.197.250
	  default-lease-time 600 max-lease-time 7200
	  option subnet-mask 255.255.255.0
	  option broadcast-address 239.252.197.255
	  option routers 239.252.197.1
	  option domain-name-servers 239.252.197.2, 239.252.197.3
	  option domain-name "isc.org";
.fi
.PP
A bootp host on that subnet that needs to be in a different domain and
use a different name server might be declared as follows:
.nf
.sp 1
	host haagen hardware ethernet 08:00:2b:4c:59:23
	  fixed-address 239.252.197.9
	  filename "/tftpboot/haagen.boot"
	  option domain-name-servers 192.5.5.1
	  option domain-name "vix.com";
.fi
.PP
A complete list of DHCP Options and their syntaxes is provided in
dhcpd.conf(5).
.SH FILES
.B ETCDIR/dhcpd.conf, DBDIR/dhcpd.leases, RUNDIR/dhcpd.pid,
.B DBDIR/dhcpd.leases~.
.SH SEE ALSO
dhcpd.conf(5), dhcpd.leases(5)
.SH AUTHOR
.B dhcpd(8)
was written by Ted Lemon <mellon@vix.com>
under a contract with Vixie Labs.   Funding
for this project was provided by the Internet Software Corporation.
Information about the Internet Software Consortium can be found at
.B http://www.isc.org/isc.


CHANGES
970607

- Change size_t to ssize_t in return values where a negative number
  is used to indicate an error.

- Always write out two digits for single-byte quantities in arrays.

- When parsing a lease database, correctly transfer the client
  hostname and hostname to the memory-resident lease structure.

970605

- Add client-hostname token to lexer so that the parser can use it.
  Fixes a serious lease database bug.

- Disable log message on receipt of short ICMP Echo replies.

970602

- Added DHCP Client scripts for FreeBSD, Solaris, and Linux, but
  they're not guaranteed to work.

- Added some Cygwin32 (Windows NT/Windows 95) support, but this is not
  sufficiently complete to be useful yet.

- Updated README

- Put something useful in TODO - formerly it mostly listed projects
  that were way out on the back burner.

In DHCP Client:

- Add default, supersede, prepend and append option support, so that a
  client can override or modify server-supplied options, and provide
  default values if the server provides no values.

- Add reject keyword, so that packets from rogue DHCP or BOOTP servers
  can be rejected out of hand.

- Added support for booting from BOOTP servers.

- Added BOOTP flag to client lease declaration, to indicated that a
  particular lease was acquired through a BOOTP server.

- Don't try to do INIT-REBOOT on leases acquired from BOOTP servers.

- Print server's IP address instead of its IP address when logging
  DHCP/BOOTP messages received by client.

- Fix some bugs in saved lease activation.

- Fix some scripting bugs.

- New sample dhclient.conf script demonstrates new features.

In common code:

- Partially implemented asynchronous DNS lookups.

- Fixed some bugs in dispatch routine.

- Fix date parsing bug that was setting dates forward one day every
  time dhcpd was restarted (this has been fixed for a while in the 1.0
  branch).

- Change name-server option name to ien116-name-server so as to reduce
  the potential for confusion.

DHCP Relay daemon:

- Fixed an operator precedence bug having to do with the broadcast
  flag.

DHCP Server:

- Add support to record the client-supplied hostname in the lease file,
  for better readability.

- Fixed a bug in the renewal code that resulted in the server ignoring
  unicast renewals from non-local subnets.   This bug caused some
  heartburn for Win95 machines.

- Copy ciaddr from saved ciaddr, not from giaddr.

- New -t flag tests /etc/dhcpd.conf for syntax errors.



dhcpd.c
/* dhcpd.c

   DHCP Server Daemon. */

/*
 * Copyright (c) 1995, 1996 The Internet Software Consortium.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of The Internet Software Consortium nor the names
 *    of its contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE INTERNET SOFTWARE CONSORTIUM AND
 * CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE INTERNET SOFTWARE CONSORTIUM OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * This software has been written for the Internet Software Consortium
 * by Ted Lemon <mellon@fugue.com> in cooperation with Vixie
 * Enterprises.  To learn more about the Internet Software Consortium,
 * see ``http://www.vix.com/isc''.  To learn more about Vixie
 * Enterprises, see ``http://www.vix.com''.
 */

static char objcopyright[] =
"$Id: dhcpd.c,v 1.30 1996/08/29 20:15:32 mellon Exp $ Copyright 1995, 1996 The Internet Software Consortium.";
static char copyright[] =
"Copyright 1995, 1996 The Internet Software Consortium.";
static char arr [] = "All rights reserved.";
static char message [] = "Internet Software Consortium DHCPD $Name:  $";

#include "dhcpd.h"

static void usage PROTO ((void));

TIME cur_time;
struct group root_group;

struct iaddr server_identifier;
int server_identifier_matched;

#ifdef USE_FALLBACK
struct interface_info fallback_interface;
#endif

u_int16_t server_port;
int log_priority;
int log_perror = 1;

int main (argc, argv, envp)
	int argc;
	char **argv, **envp;
{
	int i, status;
	struct servent *ent;
	int pidfilewritten = 0;
	char *s;
#ifndef DEBUG
	int pid;
	char pbuf [20];
	int daemon = 1;
#endif

	/* Initially, log errors to stderr as well as to syslogd. */
#ifdef SYSLOG_4_2
	openlog ("dhcpd", LOG_NDELAY);
	log_priority = DHCPD_LOG_FACILITY;
#else
	openlog ("dhcpd", LOG_NDELAY, DHCPD_LOG_FACILITY);
#endif

#ifndef DEBUG
#ifndef SYSLOG_4_2
	setlogmask (LOG_UPTO (LOG_INFO));
#endif
#endif	
	note (message);
	note (copyright);
	note (arr);

	for (i = 1; i < argc; i++) {
		if (!strcmp (argv [i], "-p")) {
			if (++i == argc)
				usage ();
			for (s = argv [i]; *s; s++)
				if (!isdigit (*s))
					error ("%s: not a valid UDP port",
					       argv [i]);
			server_port = atoi (argv [i]);
			if (server_port < 1 || server_port > 65535)
				error ("%s: not a valid UDP port",
				       argv [i]);
			server_port = htons (server_port);
			debug ("binding to user-specified port %d",
			       ntohs (server_port));
		} else if (!strcmp (argv [i], "-f")) {
#ifndef DEBUG
			daemon = 0;
#endif
		} else if (!strcmp (argv [i], "-d")) {
#ifndef DEBUG
			daemon = 0;
			log_perror = -1;
#endif
		} else if (argv [i][0] == '-') {
			usage ();
		} else {
			struct interface_info *tmp =
				((struct interface_info *)
				 dmalloc (sizeof *tmp, "get_interface_list"));
			if (!tmp)
				error ("Insufficient memory to %s %s",
				       "record interface", argv [i]);
			memset (tmp, 0, sizeof *tmp);
			strcpy (tmp -> name, argv [i]);
			tmp -> next = interfaces;
			tmp -> flags = INTERFACE_REQUESTED;
			interfaces = tmp;
		}
	}

	/* If we were requested to log to stdout on the command line,
	   keep doing so; otherwise, stop. */
	if (log_perror == -1)
		log_perror = 1;
	else
		log_perror = 0;

#ifndef DEBUG
	if (daemon) {
		/* Become a daemon... */
		if ((pid = fork ()) < 0)
			error ("Can't fork daemon: %m");
		else if (pid)
			exit (0);
		/* Become session leader and get pid... */
		pid = setsid ();
	}

	/* Read previous pid file. */
	if ((i = open (_PATH_DHCPD_PID, O_RDONLY)) >= 0) {
		status = read (i, pbuf, (sizeof pbuf) - 1);
		close (i);
		pbuf [status] = 0;
		pid = atoi (pbuf);

		/* If the previous server process is not still running,
		   write a new pid file immediately. */
		if (pid && kill (pid, 0) < 0) {
			unlink (_PATH_DHCPD_PID);
			if ((i = open (_PATH_DHCPD_PID,
				       O_WRONLY | O_CREAT, 0640)) >= 0) {
				sprintf (pbuf, "%d\n", (int)getpid ());
				write (i, pbuf, strlen (pbuf));
				close (i);
				pidfilewritten = 1;
			}
		}
	}
#endif /* !DEBUG */

	/* Default to the DHCP/BOOTP port. */
	if (!server_port)
	{
		ent = getservbyname ("dhcp", "udp");
		if (!ent)
			server_port = htons (67);
		else
			server_port = ent -> s_port;
		endservent ();
	}
  
	/* Get the current time... */
	GET_TIME (&cur_time);

	/* Read the dhcpd.conf file... */
	if (!readconf ())
		error ("Configuration file errors encountered -- exiting");

	/* Start up the database... */
	db_startup ();

	/* Discover all the network interfaces and initialize them. */
	discover_interfaces ();

#ifndef DEBUG
	/* If we didn't write the pid file earlier because we found a
	   process running the logged pid, but we made it to here,
	   meaning nothing is listening on the bootp port, then write
	   the pid file out - what's in it now is bogus anyway. */
	if (!pidfilewritten) {
		unlink (_PATH_DHCPD_PID);
		if ((i = open (_PATH_DHCPD_PID,
			       O_WRONLY | O_CREAT, 0640)) >= 0) {
			sprintf (pbuf, "%d\n", (int)getpid ());
			write (i, pbuf, strlen (pbuf));
			close (i);
			pidfilewritten = 1;
		}
	}
#endif /* !DEBUG */

	/* Receive packets and dispatch them... */
	dispatch ();

	/* Not reached */
	return 0;
}

/* Print usage message. */

static void usage ()
{
	error ("Usage: dhcpd [-p <UDP port #>] [-d] [-f] [if0 [...ifN]]");
}

void cleanup ()
{
}
